Tests/PackageGraphTests/PubgrubTests.swift

```diff
@@ -2193,7 +2193,7 @@ class DependencyGraphBuilder {
             let packageDependencies: [MockContainer.Dependency] = filteredDependencies.map {
                 (container: reference(for: $0), requirement: $1.0, products: $1.1)
             }
-            container.dependencies[version.description, default: [:]][product] = packageDependencies
+            container.dependencies[version.description, default: [:]][product, default: []] += packageDependencies
```

Comment:
I found a new problem in `serve` where even if all duplicate keys (`product`s) are preserved, each of their value (`dependencies`) is overridden by that of the next duplicate key.

I added a new commit to this PR that fixes it, reworded the previous commit, and rebased on top of the main branch.